py/src/call.py

```diff
@@ -613,26 +666,39 @@ def fetch_patricia_node(self, suffix):
         return only
 
     def fetch_contract_state(self, suffix):
+        from starkware.starkware_utils.commitment_tree.patricia_tree.patricia_tree import (
+            PatriciaTree,
+        )
+        from starkware.starknet.business_logic.fact_state.contract_state_objects import (
+            ContractState,
+        )
+
         cursor = self.connection.execute(
-            "select hash, root from contract_states where state_hash = ?", [suffix]
+            "select hash, root, nonce from contract_states where state_hash = ?", [suffix]
         )
 
-        only = next(cursor, [None, None])
+        only = next(cursor, [None, None, None])
 
-        [h, root] = only
+        [h, root, nonce] = only
 
         if h is None or root is None:
-            if suffix == b"\x00" * 32:
-                # this means that they went looking for a leaf in the patricia tree
-                # but couldn't find anything which is signalled by many zeros key
-                raise NoSuchContract
-            return None
+            # finding contract_states is a special quest.
+            #
+            # we must return this because None is not handled by caller. for
+            # some reason there is opposition to stopping cairo-lang's search
+            # when they don't find a key in the patricia tree, so it ends up
+            # making what seems like full height queries and then reads
+            # "contract_state:00..00" key, for which it hopes to find this to
+            # know that the leaf and the contract state did not exist.
+            return NOT_FOUND_CONTRACT_STATE
```

Comment:
This fix could be a standalone PR.

Reply:
Did SW ever get back to us about wtf this is about?

Reply:
> Did SW ever get back to us about wtf this is about?

No ... I just ... looked around long enough to find out that this is the expected behaviour.